### ISSUE TITLE
Placeholders for Welcome messages

### ISSUE THREAD

**Issue by realmain** [nmlorg/metabot#51]
Be able to use placeholders to call id or names within the welcome message.

* `$title` - Title of chatroom
* `$name` - Name of person
* `$username` - Username of person, auto-replace with `$name` if incoming user does not have a username

i.e. "Welcome to $title, $username!"

From **realmain** joining channel.

Output: "Welcome to GitHub, @realmain!"

**Comment by nmlorg** [comment]
One issue is that multiple people can be added at once, so `$name`, `$username`, etc. don't actually have intuitive meanings.

1. For what it's worth, [GroupButler_bot](https://t.me/GroupButler_bot) [replaces](https://github.com/group-butler/GroupButler/blob/develop/lua/groupbutler/utilities.lua#L95) the update's `from` field with the contents of the [deprecated](https://core.telegram.org/bots/api-changelog#may-18-2017) `new_chat_member` field (which only includes the first user being added), then also replaces an empty chat title with `"-"`, and applies:
   ```lua
   name = msg.from.user.first_name,
   surname = msg.from.user.last_name and msg.from.user.last_name or '',
   username = msg.from.user.username and '@'..msg.from.user.username or '-',
   id = msg.from.user.id,
   title = msg.from.chat.title,
   rules = self:deeplink_constructor(msg.from.chat.id, "rules"),
   ```
   which would be easy enough to mimic, but isn't strictly correct:
   ```
   *** User1 has added User2, User3, User4
   <bot> Welcome to GitHub, @User2!
   ```

2. Another possibility would be to build a list of all the usernames, friendly names, etc., and use [humanize.list](https://github.com/nmlorg/metabot/blob/master/metabot/util/humanize.py#L58) to format them:
   ```
   *** User1 has added User2, User3, User4
   <bot> Welcome to GitHub, @User2, @User3, and @User4!
   ```

3. A third would be to have the bot actually emit a separate greeting message for each user being added:
   ```
   *** User1 has added User2, User3, User4
   <bot> Welcome to GitHub, @User2!
   <bot> Welcome to GitHub, @User3!
   <bot> Welcome to GitHub, @User4!
   ```
I think I'd lean toward (2).

Also just double checking that you would prefer `$username` to default to `$name` instead of `"-"`, and that you wanted `$name` to include the full first/last name (`"Josh Stephens"`) instead of just first name (`"Josh"`).

**Comment by realmain** [comment]
> I think I'd lean toward (2).

Option 2 does sound best.

> Also just double checking that you would prefer `$username` to default to `$name` instead of "-", and that you wanted `$name` to include the full first/last name (`"Josh Stephens"`) instead of just first name (`"Josh"`).

I think just first name is fine actually. Unless you wanted to add both `$name` (as first name) & `$surname` placeholders for others who might want to use them.

**Comment by nmlorg** [comment]
4. Instead of simple $-substitutions, we could interpret the text as a template string (see #31), so instead of:
`Welcome to $title, $username!`
you could, assuming `update.message` is the context, do:
`Welcome to {{chat.title}}, {{new_chat_member.username and "@" + new_chat_member.username or "-"}}!`
to explicitly mimic GroupButler's behavior, or something like:
`Welcome to {{chat.title}}, {% for user in new_chat_members %}{{user.username}}{% if not loop.last %}, {% endif %}{% endfor %}!`
to approximate (2).

5. Halfway in between (2) and (4) (or maybe even in addition to (4)), go ahead and make the preprocessed/simplified variables available, just use template string syntax instead of shell syntax:
`Welcome to {{title}}, {{username}}!`
(which would expand to `"Welcome to GitHub, @User2, @User3, and @User4!"`).

I think I'm leaning toward (5) now, unless you think `{{variable}}` versus `$variable` would just be too off-putting, but it would add another external dependency ([Jinja2](https://pypi.org/project/Jinja2/), plus [MarkupSafe](https://pypi.org/project/MarkupSafe/)), and might delay things a bit to think about whether/how a shared `jinja2.Environment` instance is made available, and open a new attack surface.

**Comment by realmain** [comment]
(5) should work fine. It seems to be easy enough for the end user.

Reminder, we have to show available variables in the module for end users to know which they can use.

**Comment by nmlorg** [comment]
I've been poking around trying to find some introduction to templating (either for Jinja2 or Django) that skips over file organization, blocks, even just Python stub code, and am not really having much luck. We could just leave it as:

> You can use the following variables in your message text:
>
> {{name}} — The user's first name
> {{username}} — The user's ...
>
> For example:&nbsp; Welcome to {{title}}, {{name}}!
>
> See [Jinja2 Templates](http://jinja.pocoo.org/docs/2.10/templates/) for advanced instructions.

i.e. just the list of context variables in their simplest forms, that quick example, and a link to the Jinja2 docs, but it kinda feels like a waste given how that documentation's inaccessibility virtually guarantees nobody will ever do anything beyond using the context variables in their simplest forms.

**Comment by nmlorg** [comment]
Two quick thoughts:

We already do one substitution:&nbsp; of the naked phrase "pinned message". Would there be any merit to just replacing the naked words "name", "username", "title", etc.? And/or should the existing replacement be changed so users had to use $pinned_message or {{pinned message}}?

Several groups include bot commands in their greetings (commonly `/rules` or `/events`), and sometimes these commands are marked private-only (like multiline `/rules`). If there's a straightforward way to do it, these embedded commands should be replaced with deeplinks (instead of letting Telegram convert them to in-group bot_commands) to avoid the step of requiring users to click "Let's take this ...".

**Comment by nmlorg** [comment]
> We already do one substitution:  of the naked phrase "pinned message". Would there be any merit to just replacing the naked words "name", "username", "title", etc.? And/or should the existing replacement be changed so users had to use $pinned_message or {{pinned message}}?

Though an argument could be made that linkifying otherwise unmodified text (`AAA pinned message BBB` &rarr; `AAA <a href="https://t.me/...">pinned message</a> BBB`) is different from substituting placeholders with entirely different text (`AAA name BBB` &rarr; `AAA Josh BBB`). That might suggest that it should actually accept anything of the form `<name>`, `(name)`, `#name`, etc. in addition to `$name` and `{{name}}` (and `{name}`, and `$name$` for that matter).

Another quick thought:&nbsp; Should this actually continue to be done in [moderator.join](https://github.com/nmlorg/metabot/blob/cfb9f3145d9e51f0c1c05a5138f5d408855e0ad1/metabot/modules/moderator.py#L29) or should it actually be moved to [MessageBuilder.reply](https://github.com/nmlorg/metabot/blob/cfb9f3145d9e51f0c1c05a5138f5d408855e0ad1/metabot/util/msgbuilder.py#L72) (or even [ntelebot.preprocess.Context.reply_text](https://github.com/nmlorg/ntelebot/blob/f799dd8a4c09826ec52beafd3718f6de0f19857f/ntelebot/preprocess.py#L142)), so substitutions are basically allowed in any response sent by the bot (initially immediately including `echo` commands)?

**Comment by realmain** [comment]
> Would there be any merit to just replacing the naked words "name", "username", "title", etc.? 

That wouldn't be good because some people might actually want to use those words in the welcome message. 

> And/or should the existing replacement be changed so users had to use $pinned_message or {{pinned message}}?

That could be good too. And the bot will automatically pull the current pinned message.

**Comment by nmlorg** [comment]
> > Would there be any merit to just replacing the naked words "name", "username", "title", etc.?
> 
> That wouldn't be good because some people might actually want to use those words in the welcome message.

Right, like `Hello, my name is FunBot!` &rarr; `Hello, my Josh is FunBot!`. But we could make the keywords more obscure, like literally using `first_name`, as in `Hello, first_name, my name is FunBot!` &rarr; `Hello, Josh, my name is FunBot!`. I think I like the idea of being as permissive as possible, since most current bot admins are far from technically minded :)&nbsp; In that case, it'd be down to whether we just use unlikely keywords (like `first_name` instead of `name`) or go with the ultra-permissive delimiter approach (accepting `$name`, `{name}`, `<name>`, maybe even allow all-caps `NAME`, etc.).

Welcome to $title, $name!

Welcome to {{title}}, {{name}}!

Welcome to chat_title, first_name!

Welcome to chat title, first name!

Welcome to TITLE, NAME!
&nbsp;

> > And/or should the existing replacement be changed so users had to use $pinned_message or {{pinned message}}?
> 
> That could be good too. And the bot will automatically pull the current pinned message.

The bot already records when it sees an admin pin a message, and [replaces the string "pinned message" with a link to the pinned message](https://github.com/nmlorg/metabot/blob/cfb9f3145d9e51f0c1c05a5138f5d408855e0ad1/metabot/modules/moderator.py#L29). We can open a separate RFE to have the bot call [getChat](https://core.telegram.org/bots/api#getchat) on join to record the current pinned message.

The issue I was asking about was whether it should continue replacing the naked string "pinned message" (as in `Welcome, please read the pinned message!` &rarr; `Welcome, please read the <a href="https://t.me/groupusername/messagenumber">pinned message</a>!`) or if it should only modify text that's explicitly delimited (`Welcome, please read the $pinned_message!`). We can argue that linkifying text is different from filling in placeholders, and support `Welcome, $name, please read the pinned message!`, but otherwise it's precedent for going with `Welcome, first name, please read the pinned message!` or at least `Welcome, NAME, please read the pinned message!`.

**Comment by realmain** [comment]
I feel like it'd be safer to use `$` or `{{ }}` would be safer than a naked string to prevent unwanted replacements, including the pinned message.
